File: sns/src/main/scala/models/Message.scala
/*
 * Copyright 2012 Pellucid and Zenexity
 * 
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *   http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package aws.sns

import play.api.libs.json._

case class Message(default: String,
                   http: Option[String] = None,
                   https: Option[String] = None,
                   email: Option[String] = None,
                   emailJson: Option[String] = None,
                   sqs: Option[String] = None) {

  def serialize: String = this match {
    case Message(d, None, None, None, None, None) => d
    case Message(d, h, hs, e, ej, s) => Json.obj(
      "default" -> d,
      "http" -> h,
      "https" -> hs,
      "email" -> e,
      "emailJson" -> ej,
      "sqs" -> s).toString
  }

  def json: Boolean = this match {
    case Message(d, None, None, None, None, None) => false
    case _ => true
  }

}

File: sns/src/main/java/com/pellucid/aws/sns/Message.java
package com.pellucid.aws.sns;

import play.libs.Scala;

public class Message {

    private final String defaultValue;
    private final String http;
    private final String https;
    private final String email;
    private final String emailJson;
    private final String sas;

    public Message(String defaultValue) {
        this(defaultValue, null, null, null, null, null);
    }

    public Message(String defaultValue, String http, String https, String email, String emailJson, String sas) {
        this.defaultValue = defaultValue;
        this.http = http;
        this.https = https;
        this.email = email;
        this.emailJson = emailJson;
        this.sas = sas;
    }

    public Message withHttp(String http) {
        return new Message(
                this.defaultValue,
                http,
                this.https,
                this.email,
                this.emailJson,
                this.sas
                );
    }

    public Message withHttps(String https) {
        return new Message(
                this.defaultValue,
                this.http,
                https,
                this.email,
                this.emailJson,
                this.sas
                );
    }

    public Message withEmail(String email) {
        return new Message(
                this.defaultValue,
                this.http,
                this.https,
                email,
                this.emailJson,
                this.sas
                );
    }

    public Message withEmailJson(String emailJson) {
        return new Message(
                this.defaultValue,
                this.http,
                this.https,
                this.email,
                emailJson,
                this.sas
                );
    }

    public Message withSAS(String sas) {
        return new Message(
                this.defaultValue,
                this.http,
                this.https,
                this.email,
                this.emailJson,
                sas
                );
    }

    public aws.sns.Message toScala() {
        return new aws.sns.Message(
                defaultValue,
                Scala.Option(http),
                Scala.Option(https),
                Scala.Option(email),
                Scala.Option(emailJson),
                Scala.Option(sas)
               );
    }

}
